src/main/java/appeng/me/storage/AbstractCellInventory.java

```diff
@@ -0,0 +1,374 @@
+		this.storedItemCount = this.tagCompound.getInteger( ITEM_COUNT_TAG );
+		this.cellItems = null;
+	}
+
+	protected boolean isEmpty( final IMEInventory meInventory )
+	{
+		return meInventory.getAvailableItems( getChannel().createList() ).isEmpty();
+	}
+
+	protected IItemList<T> getCellItems()
+	{
+		if( this.cellItems == null )
+		{
+			this.cellItems = getChannel().createList();
+			this.loadCellItems();
+		}
+
+		return this.cellItems;
+	}
+
+	protected void updateItemCount( final long delta )
+	{
+		this.storedItemCount += delta;
+		this.tagCompound.setInteger( ITEM_COUNT_TAG, this.storedItemCount );
+	}
+
+	protected void saveChanges()
+	{
+		// cellItems.clean();
+		int itemCount = 0;
+
+		// add new pretty stuff...
+		int x = 0;
+		for( final T v : this.cellItems )
+		{
+			itemCount += v.getStackSize();
+
+			final NBTTagCompound g = new NBTTagCompound();
+			v.writeToNBT( g );
+			this.tagCompound.setTag( itemSlots[x], g );
+
+			this.tagCompound.setInteger( itemSlotCount[x], (int) v.getStackSize() );
+
+			x++;
+		}
+
+		// NBTBase tagType = tagCompound.getTag( ITEM_TYPE_TAG );
+		// NBTBase tagCount = tagCompound.getTag( ITEM_COUNT_TAG );
+		final short oldStoredItems = this.storedItems;
+
+		/*
+		 * if ( tagType instanceof NBTTagShort ) ((NBTTagShort) tagType).data = storedItems = (short) cellItems.size();
+		 * else
+		 */
+		this.storedItems = (short) this.cellItems.size();
+		if( this.cellItems.isEmpty() )
+		{
+			this.tagCompound.removeTag( ITEM_TYPE_TAG );
+		}
+		else
+		{
+			this.tagCompound.setShort( ITEM_TYPE_TAG, this.storedItems );
+		}
+
+		/*
+		 * if ( tagCount instanceof NBTTagInt ) ((NBTTagInt) tagCount).data = storedItemCount = itemCount; else
+		 */
+		this.storedItemCount = itemCount;
+		if( itemCount == 0 )
+		{
+			this.tagCompound.removeTag( ITEM_COUNT_TAG );
+		}
+		else
+		{
+			this.tagCompound.setInteger( ITEM_COUNT_TAG, itemCount );
+		}
+
+		// clean any old crusty stuff...
+		for( ; x < oldStoredItems && x < this.maxItemTypes; x++ )
+		{
+			this.tagCompound.removeTag( itemSlots[x] );
+			this.tagCompound.removeTag( itemSlotCount[x] );
+		}
+
+		if( this.container != null )
+		{
+			this.container.saveChanges( this );
+		}
+	}
+
+	private void loadCellItems()
+	{
+		if( this.cellItems == null )
+		{
+			this.cellItems = getChannel().createList();
+		}
+
+		this.cellItems.resetStatus(); // clears totals and stuff.
+
+		final int types = (int) this.getStoredItemTypes();
+
+		for( int slot = 0; slot < types; slot++ )
+		{
+			NBTTagCompound compoundTag = this.tagCompound.getCompoundTag( itemSlots[slot] );
+			int stackSize = this.tagCompound.getInteger( itemSlotCount[slot] );
+			this.loadCellItem( compoundTag, stackSize );
+		}
+
+		// cellItems.clean();
+	}
+
+	protected abstract void loadCellItem( NBTTagCompound compoundTag, int stackSize );
+
+	@Override
+	public IItemList getAvailableItems( final IItemList out )
+	{
+		for( final T i : this.getCellItems() )
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Rename "i" which hides the field declared at line 62. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.orod.org/coding_rules#rule_key=squid%3AHiddenFieldCheck)